### PR TITLE
Contribution Guidelines

### PR DESCRIPTION
**Pull request by maxfischer2781**
Adds a contribution guideline and establishes automatic linter, unittest and coverage runs.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Contributing to μSim
+
+μSim is hosted on GitHub and open source.
+This makes it easy for you to submit bug reports or directly propose changes.
+
+## Reporting a Bug
+
+Enabling precise and reliable simulations is our goal,
+and that means swatting every single bug.
+However, there is only so much time in a day -
```
Comment (eileen-kuehn) [inline CONTRIBUTING.md]:
```suggestion
However, there is so little time in a day -
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Contributing to μSim
+
+μSim is hosted on GitHub and open source.
+This makes it easy for you to submit bug reports or directly propose changes.
+
+## Reporting a Bug
+
+Enabling precise and reliable simulations is our goal,
+and that means swatting every single bug.
```
Comment (eileen-kuehn) [inline CONTRIBUTING.md]:
```suggestion
and that means chasing every single bug.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+However, there is only so much time in a day -
+Please help us help you!
+
+* Avoid reporting issues that are already being worked on:
+  * *Exclude version issues:*
+    Test whether the issue also occurs with the most recent version of μSim.
+  * *Check previous reports:*
+    Have a look at [previous reports](https://github.com/MaineKuehn/usim/issues?q=is%3Aissue+label%3Abug)
+    if they match your issue.
```
Comment (eileen-kuehn) [inline CONTRIBUTING.md]:
```suggestion
    if any of those matches your issue.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  * Describe both the *desired* behaviour you expected,
+    as well as *actual* behaviour you observed.
+  * Provide basic code that reproduces or at least illustrates your setup.
+    If the error caused an Exception, include the traceback.
+
+## Submitting Fixes and Features
+
+We welcome direct contributions of code and documentation,
+but also need to maintain the current and future quality of μSim.
+The better you integrate with the development of μSim,
```
Comment (eileen-kuehn) [inline CONTRIBUTING.md]:
```suggestion
The better you integrate with the development style of μSim,
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+the more likely and faster can your contribution be included.
```
Comment (eileen-kuehn) [inline CONTRIBUTING.md]:
```suggestion
the more likely and faster we can include your contribution.
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    If there is no open issue yet, it might be a good idea to create one for documentation.
+  * Check the [open pull requests](https://github.com/MaineKuehn/usim/pulls)
+    if they are similar to your changes.
+    Make sure to assess how your contribution relates to conflicting changes.
+* Open a [pull request](https://github.com/MaineKuehn/usim/pulls) for your changes:
+  * Use a title that describes the contribution as clearly as possible.
+    For new features, prepend ``WIP:`` to the title to show that it needs discussion.
+  * Provide a description of the changes, ideally with a short example.
+    Include references to all related issues and pull requests.
+* Your changes will be reviewed before being merged
```
Comment (eileen-kuehn) [inline CONTRIBUTING.md]:
```suggestion
* We will review your changes before merging
```

File: usim/_core/loop.py [hunk 7]
```diff
@@ -2,8 +2,10 @@
 event handling core facilities
 
 This is the core scheduling/synchronization implementation.
-It defines a basic event-loop with simulated time, as well as the signals understood by the loop.
-The entirety of this module is for internal use only, and should not be used by simulation code directly.
+It defines a basic event-loop with simulated time, as well as the signals
```
Comment (eileen-kuehn) [inline usim/_core/loop.py]:
```suggestion
It defines a basic event loop with simulated time, as well as the signals
```

File: usim/_core/loop.py [hunk 8]
```diff
@@ -66,13 +72,16 @@ class Loop:
...
+    coroutines are synchronized.
     Commands are either ``await``\ able, issued in an ``async def`` coroutine,
-    or side-channel via ``__LOOP_STATE__.LOOP``, issued by either coroutines or subroutines.
+    or side-channel via ``__LOOP_STATE__.LOOP``,
+    issued by either coroutines or subroutines.
 
     The event loop understands only a single ``await``\ able command,
     :py:class:`Hibernate`, which immediately suspends the currently active coroutine.
-    This does *not* schedule the coroutine for resumption, which must be done separately.
+    This does *not* schedule the coroutine for resumption,
```
Comment (eileen-kuehn) [inline usim/_core/loop.py]:
```suggestion
    This does *not* schedule the coroutine for resumption.
```

File: usim/_core/loop.py [hunk 9]
```diff
@@ -66,13 +72,16 @@ class Loop:
...
+    which must be done separately.
```
Comment (eileen-kuehn) [inline usim/_core/loop.py]:
```suggestion
    This must be done separately.
```

File: usim/_core/loop.py [hunk 10]
```diff
@@ -168,7 +183,10 @@ def _run_coroutine(self, target: Coroutine, signal: BaseException = None):
         except BaseException as err:
             raise ActivityError(target, signal=signal) from err
 
-    def schedule(self, target: Coroutine, signal: 'Interrupt' = None, *, delay: float = None, at: float = None):
+    def schedule(
+            self, target: Coroutine, signal: 'Interrupt' = None, *,
```
Comment (eileen-kuehn) [inline usim/_core/loop.py]:
```suggestion
            self, 
            target: Coroutine, 
            signal: 'Interrupt' = None, *,
```

File: usim/_core/loop.py [hunk 11]
```diff
@@ -168,7 +183,10 @@ def _run_coroutine(self, target: Coroutine, signal: BaseException = None):
...
+            delay: float = None, at: float = None
```
Comment (eileen-kuehn) [inline usim/_core/loop.py]:
```suggestion
            delay: float = None, 
            at: float = None
```

File: usim/_primitives/condition.py [hunk 12]
```diff
@@ -97,8 +98,13 @@ def __init__(self, *conditions: Condition):
...
         self._children = tuple(
-            set(condition for condition in conditions if not isinstance(condition, self.__class__)) |
-            set(condition._children for condition in conditions if isinstance(condition, self.__class__))
+            set(condition for condition in conditions if not isinstance(
+                condition, self.__class__
+            )).union(
+                condition._children for condition in conditions if isinstance(
+                    condition, self.__class__
+                )
+            )
```
Comment (eileen-kuehn) [inline usim/_primitives/condition.py]: This is very bad maintainable. We should consider for refactoring.

File: usim/_primitives/context.py [hunk 13]
```diff
@@ -163,8 +174,10 @@ def until(notification: Notification):
     r"""
     :py:class:`Scope` that is interrupted on notification
 
-    An asynchronous `until`-scope listens for a notification *without* stopping execution.
-    This allows notification on any break point, i.e. `await` in the context or while waiting for children.
+    An asynchronous `until`-scope listens for a notification
+    *without* stopping execution.
+    This allows notification on any break point,
+    i.e. `await` in the context or while waiting for children.
```
Comment (eileen-kuehn) [inline usim/_primitives/context.py]:
```suggestion
    e.g. `await` in the context or while waiting for children.
```

File: usim/_primitives/task.py [hunk 14]
```diff
@@ -60,10 +60,13 @@ class Task(Awaitable[RT]):
...
 
-    A :py:class:`Task` wraps an :term:`activity` that is concurrently run in a :py:class:`~.Scope`.
-    This allows to store or pass on the :py:class:`Task` in order to control the underlying activity.
+    A :py:class:`Task` wraps an :term:`activity` that is
+    concurrently run in a :py:class:`~.Scope`.
+    This allows to store or pass on the :py:class:`Task`
+    in order to control the underlying activity.
     Other activities can ``await`` a :py:class:`Task`
-    to receive any results or exceptions on completion, similar to a regular activity.
+    to receive any results or exceptions on completion,
```
Comment (eileen-kuehn) [inline usim/_primitives/task.py]:
```suggestion
    to receive results or exceptions on completion,
```

File: usim/_primitives/timing.py [hunk 15]
```diff
@@ -1,8 +1,10 @@
 """
-There is no inherent time unit, such as seconds, hours, or years, implied for simulation time.
+There is no inherent time unit, such as seconds, hours, or years,
```
Comment (eileen-kuehn) [inline usim/_primitives/timing.py]:
```suggestion
For simulation time there is no inherent time unit, such as seconds,
```

File: usim/_primitives/timing.py [hunk 16]
```diff
@@ -1,8 +1,10 @@
...
+implied for simulation time.
```
Comment (eileen-kuehn) [inline usim/_primitives/timing.py]:
```suggestion
hours, or years, implied.
```

File: usim/_primitives/timing.py [hunk 17]
```diff
@@ -1,8 +1,10 @@
...
 A simulation should use a consistent time unit, however.
 
-The default time type is :py:class:`float`, which in principle may exhibit imprecision for fractions.
+The default time type is :py:class:`float`, which in principle may exhibit
```
Comment (eileen-kuehn) [inline usim/_primitives/timing.py]: I am struggling with *in principle*. We might write "which in theory may" or "which might".

File: usim/_primitives/condition.py (hunk 12 above)
Comment (maxfischer2781) [reply usim/_primitives/condition.py]: This actually has a bug as well. :D See issue #5, I'll open a pull request once the CI is in master.

File: usim/_core/loop.py [hunk 18]
```diff
@@ -168,7 +183,13 @@ def _run_coroutine(self, target: Coroutine, signal: BaseException = None):
         except BaseException as err:
             raise ActivityError(target, signal=signal) from err
 
-    def schedule(self, target: Coroutine, signal: 'Interrupt' = None, *, delay: float = None, at: float = None):
+    def schedule(
+            self, 
+            target: Coroutine,
```
Comment (maxfischer2781) [inline usim/_core/loop.py]:
```suggestion
            target: Coroutine,
```

File: usim/_core/loop.py [hunk 19]
```diff
@@ -168,7 +183,13 @@ def _run_coroutine(self, target: Coroutine, signal: BaseException = None):
         except BaseException as err:
             raise ActivityError(target, signal=signal) from err
 
-    def schedule(self, target: Coroutine, signal: 'Interrupt' = None, *, delay: float = None, at: float = None):
+    def schedule(
+            self, 
+            target: Coroutine, 
+            signal: 'Interrupt' = None, *,
+            delay: float = None,
```
Comment (maxfischer2781) [inline usim/_core/loop.py]:
```suggestion
            delay: float = None,
```

File: usim/_core/loop.py [hunk 20]
```diff
@@ -168,7 +183,13 @@ def _run_coroutine(self, target: Coroutine, signal: BaseException = None):
         except BaseException as err:
             raise ActivityError(target, signal=signal) from err
 
-    def schedule(self, target: Coroutine, signal: 'Interrupt' = None, *, delay: float = None, at: float = None):
+    def schedule(
+            self,
```
Comment (maxfischer2781) [inline usim/_core/loop.py]:
```suggestion
            self,
```

**Pull request merged**